### ISSUE TITLE
Got 'Undefined variable: sam' when using `g:vader_exception` right after `AssertThrows` 

### ISSUE THREAD

**Issue by cgsdfc** [junegunn/vader.vim#160]
Hi folks.
I got this suspicious bug when doing a  test. When I passed a variable `sam` defined within an `Execute` block to a function and tested the call with an `AssertThrows` and then wanted to verify the throw with `g:vader_exception`, Vader/Vim gave this error:
```vader
    (2/2) [EXECUTE] Undefined sam
      > Vim(call):E121: Undefined variable: sam
```
Here is a minimal example that repeats the situation:
```vader
Execute(Undefined sam):
  function! F(x)
    throw 'Hello'
  endfunction
  
  let sam = {}
  AssertThrows call F(sam)
  Log g:vader_exception
```
By the way, the variable `g:vader_throwpoint` does not subject to this bug.
```vader
Execute(Undefined sam):
  function! F(x)
    throw 'Hello'
  endfunction
  
  let sam = {}
  AssertThrows call F(sam)
  Log g:vader_throwpoint
```
shows:
```
    (1/1) [EXECUTE] Undefined sam
      > function vader#assert#throws, line 5
```

Have I been using the ``AssertThrows`` tool in the wrong way? Note, even if the throwing function is autoloaded, the bug is still here.

**Comment by cgsdfc** [comment]
Update: It works if I pass literal value to the function that throws, i.e., ``1``, ``[]`` or ``{}``.
```

Execute(Works fine with literal value):
  function! F(x)
    throw 'Hello'
  endfunction
  
  AssertThrows call F(1)
  Log g:vader_exception
```

**Comment by blueyed** [comment]
Noticed this myself already before, but have not investigated.
IIRC using a script-local variable `s:foo` (maybe only with #107?) might work.

**Comment by cgsdfc** [comment]
No really. I add ``s:`` to ``sam`` but the error is still here.
```
Execute(Undefined sam):
  function! F(x)
    throw 'Hello'
  endfunction
  let s:sam = {}
  AssertThrows call F(s:sam)
  Log g:vader_exception
  ```
Output:
```
Starting Vader: 1 suite(s), 2 case(s)
  Starting Vader: /home/cgsdfc/vim-object/bugs/UndefinedSam.vader
    (1/2) [EXECUTE] Undefined sam
      > Vim(call):E121: Undefined variable: s:sam
    (2/2) [EXECUTE] Works fine with literal value
      > Hello
  Success/Total: 2/2
Success/Total: 2/2 (assertions: 2/2)
Elapsed time: 0.062485 sec.
```
Currently I don't have much knowledge of the internal of ``Vader`` . I will be happy to see a fix to this as long as it is considered a bug.

**Comment by cgsdfc** [comment]
As a simple work-around, you can verify the exception thrown is as expected. 
```
Execute(Work around the 'Undefined Sam');
  function! F(x)
    throw 'Hello'
  endfunction
  
  let sam = {}
  try
    call F(sam)
  catch
    Assert v:exception =~# 'Hello'
  endtry
```
Output:
```
Starting Vader: 1 suite(s), 1 case(s)
  Starting Vader: /home/cgsdfc/vim-object/bugs/workaround.vader
    (1/1) [EXECUTE] Work around the 'Undefined Sam'
  Success/Total: 1/1
Success/Total: 1/1 (assertions: 1/1)
Elapsed time: 0.057876 sec.
```
If you only care about an exception is thrown, use ``AssertThrows`` instead.

**Comment by cgsdfc** [comment]
I get some insight of this problem. The problem is about the change of default scope when unqualified names are looked up. To eliminate this bug, please explicitly prefix your variable with ``g:`` in the ``AssertThrows`` command.
```
AssertThrows call F(g:sam)
```
For the complete reasons, please watch the following code:
```vim
" The term 'default scope' refers the the scope where the
" unqualified names are looked up, i.e., x, y.

" This shows that default scope in function is l:.
function! F()
  let x = 1
  execute 'echo l:x is# x'
endfunction

" This explains the root of the bug.
function! X()
  " the l: hides the g:
  execute 'echo x'
endfunction

" Use this as a solution.
function! G()
  " Explicitly prefix a g:.
  execute 'echo g:x'
endfunction

function! W()
  " This explains why s: prefix does not work.
  " s: variable can only be used in a _script_, not in Ex mode
  " and thus not in execute command since it is a shortcut for
  " the Ex mode.
  let s:x = 1
  " Illformed, no s: is allowed.
  execute 'echo s:x'
endfunction

call F()
call X()
call G()
call W()

" In the default scope of file level is g:.
let x = 0
echo 'default scope is g: ?' x is# g:x ? 'yes' : 'no'
```
The knowledge from the code base help me. 

* vader/assert.vim tells me ``AssertThrows`` is done via ``execute`` inside a function.
* vader/window.vim tells me ``Execute`` block is first write to a temp file which will be sourced. Thus, all the unqualified variables is effectively inside ``g:``.

In short, this problem is due completely to the scoping rules of vim, which is relatively hard to fix if possible. I hence suggest an update to the limitation part of README that tells people to use ``g:`` in ``AssertThrows`` rather than happily and blindly believe that ``AssertThrows`` does what they believe.
